// dcorejs-sdk/dist/models/ContentTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ApplicationType;
(function (ApplicationType) {
    ApplicationType[ApplicationType["DecentCore"] = 0] = "DecentCore";
    ApplicationType[ApplicationType["DecentGo"] = 1] = "DecentGo";
    ApplicationType[ApplicationType["Alax"] = 2] = "Alax";
})(ApplicationType = exports.ApplicationType || (exports.ApplicationType = {}));
var CategoryType;
(function (CategoryType) {
    CategoryType[CategoryType["None"] = 0] = "None";
    CategoryType[CategoryType["Music"] = 1] = "Music";
    CategoryType[CategoryType["Movie"] = 2] = "Movie";
    CategoryType[CategoryType["Book"] = 3] = "Book";
    CategoryType[CategoryType["AudioBook"] = 4] = "AudioBook";
    CategoryType[CategoryType["Software"] = 5] = "Software";
    CategoryType[CategoryType["Game"] = 6] = "Game";
    CategoryType[CategoryType["Picture"] = 7] = "Picture";
    CategoryType[CategoryType["Document"] = 8] = "Document";
})(CategoryType = exports.CategoryType || (exports.CategoryType = {}));
/**
 * create content type string
 */
function contentType(app, category) {
    return app + "." + category + ".0";
}
exports.contentType = contentType;
