// dcorejs-sdk/dist/net/models/request/BaseRequest.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var class_transformer_1 = require("class-transformer");
var ApiGroup_1 = require("../ApiGroup");
var RequestJson_1 = require("../RequestJson");
var BaseRequest = /** @class */ (function () {
    function BaseRequest(apiGroup, method, params, transformer) {
        if (transformer === void 0) { transformer = function (value) { return value; }; }
        this.params = [];
        this.jsonrpc = "2.0";
        this.id = 1;
        this.apiGroup = apiGroup;
        this.transformer = transformer;
        this.method = method;
        this.params = params;
    }
    BaseRequest.prototype.description = function () {
        return "method: " + this.method + " params: " + this.params;
    };
    BaseRequest.prototype.json = function (callId, callbackId) {
        if (callId === void 0) { callId = 1; }
        return class_transformer_1.serialize(new RequestJson_1.RequestJson(this, callId, callbackId));
    };
    __decorate([
        class_transformer_1.Exclude(),
        __metadata("design:type", Number)
    ], BaseRequest.prototype, "apiGroup", void 0);
    __decorate([
        class_transformer_1.Exclude(),
        __metadata("design:type", Function)
    ], BaseRequest.prototype, "transformer", void 0);
    __decorate([
        class_transformer_1.Expose({ name: "method" }),
        __metadata("design:type", String)
    ], BaseRequest.prototype, "method", void 0);
    __decorate([
        class_transformer_1.Expose({ name: "params" }),
        __metadata("design:type", Array)
    ], BaseRequest.prototype, "params", void 0);
    __decorate([
        class_transformer_1.Expose({ name: "jsonrpc" }),
        __metadata("design:type", String)
    ], BaseRequest.prototype, "jsonrpc", void 0);
    __decorate([
        class_transformer_1.Expose({ name: "id" }),
        __metadata("design:type", Number)
    ], BaseRequest.prototype, "id", void 0);
    return BaseRequest;
}());
exports.BaseRequest = BaseRequest;
