// dcorejs-sdk/dist/DCoreApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AccountApi_1 = require("./api/AccountApi");
var AssetApi_1 = require("./api/AssetApi");
var AuthorityApi_1 = require("./api/AuthorityApi");
var BalanceApi_1 = require("./api/BalanceApi");
var BlockApi_1 = require("./api/BlockApi");
var BroadcastApi_1 = require("./api/BroadcastApi");
var CallbackApi_1 = require("./api/CallbackApi");
var ContentApi_1 = require("./api/ContentApi");
var GeneralApi_1 = require("./api/GeneralApi");
var HistoryApi_1 = require("./api/HistoryApi");
var MessageApi_1 = require("./api/MessageApi");
var MiningApi_1 = require("./api/MiningApi");
var PurchaseApi_1 = require("./api/PurchaseApi");
var SeedersApi_1 = require("./api/SeedersApi");
var SubscriptionApi_1 = require("./api/SubscriptionApi");
var TransactionApi_1 = require("./api/TransactionApi");
var ValidationApi_1 = require("./api/ValidationApi");
var DCoreConstants_1 = require("./DCoreConstants");
var DCoreApi = /** @class */ (function () {
    function DCoreApi(core) {
        this.core = core;
        this.transactionExpiration = DCoreConstants_1.DCoreConstants.EXPIRATION_DEFAULT;
        this.accountApi = new AccountApi_1.AccountApi(this);
        this.assetApi = new AssetApi_1.AssetApi(this);
        this.authorityApi = new AuthorityApi_1.AuthorityApi(this);
        this.balanceApi = new BalanceApi_1.BalanceApi(this);
        this.blockApi = new BlockApi_1.BlockApi(this);
        this.broadcastApi = new BroadcastApi_1.BroadcastApi(this);
        this.callbackApi = new CallbackApi_1.CallbackApi(this);
        this.contentApi = new ContentApi_1.ContentApi(this);
        this.generalApi = new GeneralApi_1.GeneralApi(this);
        this.historyApi = new HistoryApi_1.HistoryApi(this);
        this.messageApi = new MessageApi_1.MessageApi(this);
        this.miningApi = new MiningApi_1.MiningApi(this);
        this.purchaseApi = new PurchaseApi_1.PurchaseApi(this);
        this.seedersApi = new SeedersApi_1.SeedersApi(this);
        this.subscriptionApi = new SubscriptionApi_1.SubscriptionApi(this);
        this.transactionApi = new TransactionApi_1.TransactionApi(this, this.core);
        this.validationApi = new ValidationApi_1.ValidationApi(this);
    }
    DCoreApi.prototype.request = function (request) {
        return this.core.request(request);
    };
    DCoreApi.prototype.requestStream = function (request) {
        return this.core.requestStream(request);
    };
    DCoreApi.prototype.disconnect = function () {
        this.core.disconnect();
    };
    Object.defineProperty(DCoreApi.prototype, "timeout", {
        set: function (millis) {
            this.core.timeout = millis;
        },
        enumerable: true,
        configurable: true
    });
    return DCoreApi;
}());
exports.DCoreApi = DCoreApi;
