// dcorejs-sdk/dist/models/order/SearchAccountHistoryOrder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var SearchAccountHistoryOrder;
(function (SearchAccountHistoryOrder) {
    SearchAccountHistoryOrder["TypeAsc"] = "+type";
    SearchAccountHistoryOrder["ToAsc"] = "+to";
    SearchAccountHistoryOrder["FromAsc"] = "+from";
    SearchAccountHistoryOrder["PriceAsc"] = "+price";
    SearchAccountHistoryOrder["FeeAsc"] = "+fee";
    SearchAccountHistoryOrder["DescriptionAsc"] = "+description";
    SearchAccountHistoryOrder["TimeAsc"] = "+time";
    SearchAccountHistoryOrder["TypeDesc"] = "-type";
    SearchAccountHistoryOrder["ToDesc"] = "-to";
    SearchAccountHistoryOrder["FromDesc"] = "-from";
    SearchAccountHistoryOrder["PriceDesc"] = "-price";
    SearchAccountHistoryOrder["FeeDesc"] = "-fee";
    SearchAccountHistoryOrder["DescriptionDesc"] = "-description";
    SearchAccountHistoryOrder["TimeDesc"] = "-time";
})(SearchAccountHistoryOrder = exports.SearchAccountHistoryOrder || (exports.SearchAccountHistoryOrder = {}));
